### PR TITLE
[persist] Write fewer, larger batches in the distributed persist sink

### PR DESCRIPTION
**Pull request by bkirwi**
This branch implements a mechanism for `write_batches` to send data along to `append_batches` when the batch it would otherwise write is too small.

### Motivation

Known-desireable improvement: https://github.com/MaterializeInc/materialize/issues/15985

### Tips for reviewer

Right now this code is very conservative about when it considers a batch too small:
- I've set a pretty low threshold as a constant. (We could make this configurable if you like.)
- I'm checking the size of `correction`, not `to_append`, to avoid iterating through everything twice. The theory is that `correction` is probably pretty close in size to `to_append` in normal operation, when emitting small batches is frequent, and anyways we'll err on the side of the status quo.

Feel free to tell me if either of those doesn't feel appropriate.

I don't yet have a good way to measure the impact of this change. One possibility is to add a batch-size histogram and see whether the distribution changes. (This would also be a good way to choose a threshold.) In theory it should result in fewer parts going to compaction, etc., but I'm not sure how large of an effect that will be.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/compute/src/sink/persist_sink.rs
Comment (teskje) [inline src/compute/src/sink/persist_sink.rs:804]:
```suggestion
                        let batch_or_data = if correction.len() >= MINIMUM_BATCH_SIZE {
```

... right?

**Review by teskje — CHANGES_REQUESTED** [review]
This looks good to me, except that I think a very central comparison operator is flipped :)

I agree that it would be good if we had some persist metric to measure the impact of this change with.

File: src/compute/src/sink/persist_sink.rs
Comment (bkirwi) [reply src/compute/src/sink/persist_sink.rs:804]: 🤦‍♂️ - thank you!

**Review by teskje — APPROVED** [review]

**Comment by bkirwi** [comment]
The nightly run on my original commit passed except for the feature benchmark, which showed serious regressions in a number of benchmarks, presumably because of my dumb bug.

I'm rerunning just the feature benchmark now. I'm unsure if those benchmarks are enough to show significant improvements, but at least we'll be able to check if I'm still making things worse.

File: src/compute/src/sink/persist_sink.rs
Comment (danhhz) [inline src/compute/src/sink/persist_sink.rs:1092]: let's toss a consolidate in here, at least until we can revisit paul's change to consolidate in user batches

File: src/compute/src/sink/persist_sink.rs
Comment (danhhz) [inline src/compute/src/sink/persist_sink.rs:1094]: we could also put it in a BatchBuilder as we go, which shouldn't really have any overhead (though I'd sanity check this quickly before committing to it). if not, at least leave this as a TODO, please

**Review by danhhz — APPROVED** [review]
amazing, so excited for this!! it's the last known blocked for turning persisted introspection shards back on. LGTM modulo the PersistConfig comment

the reason I haven't yet typed this up myself was that it increases the worker imbalance of persist_sink (the single worker last operator now does more work than before) and I wasn't sure how Frank would feel about that. we should probably sanity check with him before merging unless Jan feels his LGTM is a reasonable proxy for Frank's

super nit: I'd s/larger/fewer/ in the PR title

File: src/compute/src/sink/persist_sink.rs
Comment (bkirwi) [reply src/compute/src/sink/persist_sink.rs:509]: Yeah, apologies - thought you might prefer this but forgot to mention it in the description. Happy to!

**Comment by teskje** [comment]
> unless Jan feels his LGTM is a reasonable proxy for Frank's

I absolutely don't think this!

My 2 cents would be that if we have many sink dataflows, different workers are likely to host the append_batches operator, so the load should even out. That's not the case if we have only one or a few sink dataflows, but hopefully the skew is not so bad since we would only write small batches there anyway.

**Comment by bkirwi** [comment]
Yeah, that's my feeling as well... for very small batches this shouldn't really be increasing load on any particular worker by any noticeable amount, so this is more of a tuning question than a binary one. Currently planning on using Persist's upcoming dynamic config for this, so we can merge with a very small threshold and then tune it upwards from there.

**Comment by bkirwi** [comment]
Feature benchmark is green - I'll come back to this in a bit with dynamic config and (perhaps) metrics.

File: src/compute/src/sink/persist_sink.rs
Comment (bkirwi) [reply src/compute/src/sink/persist_sink.rs]: Done; seems AOK. (Though it does prevent us from consolidating here...)

**Comment by aljoscha** [comment]
It's nice how neatly this fits in! 🎉

**Comment by danhhz** [comment]
FYI #16797 just merged if you want to pull it in and switch to a dynamic config

**Review by umanwizard — COMMENTED** [review]
The idea seems legit to me (and I'm super happy we'll be able to selectively turn on introspection sources!)

Would you mind adding a comment somewhere explaining why we're doing this? (I.e., that `active_worker` is effectively single-threaded, and why that matters)

File: src/adapter/src/session/vars.rs
Comment (danhhz) [inline src/adapter/src/session/vars.rs:381]: nit: copy-pasta

File: src/adapter/src/session/vars.rs
Comment (danhhz) [inline src/adapter/src/session/vars.rs:1452]: nit: copy-pasta

File: src/persist-client/src/cfg.rs
Comment (danhhz) [inline src/persist-client/src/cfg.rs:181]: nit: copy-pasta

File: src/persist-client/src/cfg.rs
Comment (danhhz) [inline src/persist-client/src/cfg.rs:182]: perfect! I was going to suggest 0 as the default (for the initial merge)

**Review by danhhz — APPROVED** [review]

File: src/adapter/src/session/vars.rs
Comment (bkirwi) [reply src/adapter/src/session/vars.rs:381]: 🤦

File: src/adapter/src/session/vars.rs
Comment (bkirwi) [reply src/adapter/src/session/vars.rs:1452]: 🤦

File: src/persist-client/src/cfg.rs
Comment (bkirwi) [reply src/persist-client/src/cfg.rs:181]: 🤦

**Comment by bkirwi** [comment]
A chart of puts per second when I was playing with this in staging yesterday afternoon, with persisted introspection turned on:

image

The dips are with a threshold of 16, 10k, and 128, respectively. So this has some effect, though it takes a somewhat higher threshold than I expected to really kick into gear, at least on my little test setup.

**Review by umanwizard — APPROVED** [review]

**Comment by bkirwi** [comment]
Added a couple metrics as requested - a couple counters and a histogram for the overall blob size distribution. This required a little rummaging around in `ore`, to support vanilla `Histogram`s and lossy casts. Sorry about that!

Also tagging @frankmcsherry as a reviewer, since I think he wanted to take a look at this one before it makes its way in.

**Review by benesch — APPROVED** [review]
`CastLossy` LGTM! Did not review the rest.

**Comment by bkirwi** [comment]
@frankmcsherry has elsewhere said that he's happy if we're comfortable with the details of -

> [...] what happens when different workers reach different conclusions about the size of data that they are reading, whether there is a hard cut-over from one worker to all workers, that sort of thing.

I feel like we've hashed these concerns out successfully on this PR -- thanks everyone who reviewed! --so I think this is ready to go. Will merge tomorrow unless someone objects.

**Pull request merged**
